File: src/main/java/lmller/github/io/gofkotlin/interpreter/Interpreter.java
package lmller.github.io.gofkotlin.interpreter;

public class Interpreter {
    public interface Expression {
        int eval();
    }

    public static class Add implements Expression {

        final Expression leftHand;
        final Expression rightHand;

        public Add(Expression leftHand, Expression rightHand) {
            this.leftHand = leftHand;
            this.rightHand = rightHand;
        }

        @Override
        public int eval() {
            return leftHand.eval() + rightHand.eval();
        }

        @Override
        public String toString() {
            return leftHand.toString() +" + " + rightHand.toString();
        }
    }

    public static class Subtract implements Expression {

        final Expression leftHand;
        final Expression rightHand;

        public Subtract(Expression leftHand, Expression rightHand) {
            this.leftHand = leftHand;
            this.rightHand = rightHand;
        }

        @Override
        public int eval() {
            return leftHand.eval() - rightHand.eval();
        }
        @Override
        public String toString() {
            return leftHand.toString() +" - " + rightHand.toString();
        }

    }

    public static class Number implements Expression {
        final int value;

        public Number(int value) {
            this.value = value;
        }

        @Override
        public int eval() {
            return value;
        }

        @Override
        public String toString() {
            return Integer.toString(value);
        }
    }

    public static void main(String[] args) {
        //this expression would normally be parsed, but I'm too lazy to write a parser
        Expression expression = new Add(new Number(40), new Subtract(new Number(10), new Number(8)));

        printExpression(expression);
    }


    public static void printExpression(Expression expression) {
        System.out.println(expression.toString() + " = " + expression.eval());
    }

}

File: src/main/java/lmller/github/io/gofkotlin/interpreter/Interpreter.kt
package lmller.github.io.gofkotlin.interpreter

sealed class Expression {
    abstract fun eval(): Int

    //personally, I'd still put this into the concrete classes.
    //however, this is nice to show the `when`-expression in combination with `sealed` classes
    //that could be used inside the Expression parser
    override fun toString() = when(this) {
        is Add -> "$leftHand + $rightHand"
        is Subtract -> "$leftHand - $rightHand"
        is Number -> "$value"
    }
}

class Add(val leftHand: Expression, val rightHand: Expression) : Expression() {
    override fun eval() = leftHand.eval() + rightHand.eval()
}

class Subtract(val leftHand: Expression, val rightHand: Expression) : Expression() {
    override fun eval() = leftHand.eval() - rightHand.eval()
}

class Number(val value: Int) : Expression() {
    override fun eval() = value
}

fun main(args: Array<String>) {
    //this expression would normally be parsed, but I'm too lazy to write a parser
    val expression = Add(Number(40), Subtract(Number(10), Number(8)))

    printExpression(expression)
}


fun printExpression(expression: Expression) {
    println(expression.toString() + " = " + expression.eval())
}